src/main/scala/edu/ie3/util/osm/model/OsmEntity.scala

```diff
@@ -0,0 +1,162 @@
+        true
+      case None => false
+    }
+
+  @tailrec
+  final def hasKeysValuesPairOr(
+      keyTagValues: Map[String, Set[String]]
+  ): Boolean = {
+    keyTagValues.headOption match {
+      case Some((key, tagValues)) =>
+        // if we have at least one matching key/tagValues pair
+        // we can stop the process and return true
+        if (hasKeyValuesPairOr(key, tagValues)) {
+          true
+        } else {
+          hasKeysValuesPairOr(keyTagValues - key)
+        }
+      case None =>
+        false
+    }
+  }
+
+  def hasKey(key: CommonOsmKey): Boolean =
+    hasKeyValuesPairOr(key, Set.empty)
+
+  def hasKey(key: String): Boolean =
+    hasKeyValuesPairOr(key, Set.empty)
+
+}
+
+object OsmEntity {
+
+  enum OsmEntityType:
+    case Node, Way, Relation
+
+  final case class MetaInformation(
+      version: Option[Int] = None,
+      timestamp: Option[Instant] = None,
+      changeSet: Option[Long] = None,
+      userId: Option[Int] = None,
+      userName: Option[String] = None,
+      visible: Option[Boolean] = None
+  )
+
+  final case class Node(
+      id: Long,
+      latitude: Double,
+      longitude: Double,
+      tags: Map[String, String],
+      metaInformation: Option[MetaInformation] = None
+  ) extends OsmEntity {
+    override val osmModel: OsmEntityType = OsmEntityType.Node
+    val coord: Point = GeoUtils.xyToPoint(longitude, latitude)
+  }
+
+  sealed trait Way extends OsmEntity {
+    val id: Long
+    val nodes: Seq[Long]
+    val tags: Map[String, String]
+    val metaInformation: Option[MetaInformation]
+
+    override val osmModel: OsmEntityType = OsmEntityType.Way
+  }
+
+  object Way {
+
+    final case class OpenWay private[model] (
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ) extends Way
+
+    final case class ClosedWay private[model] (
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ) extends Way
+
+    def apply(
+        id: Long,
+        nodes: Seq[Long],
+        tags: Map[String, String],
+        metaInformation: Option[MetaInformation]
+    ): Way =
+      if (closedWay(nodes)) {
+        ClosedWay(id, nodes, tags, metaInformation)
+      } else {
+        OpenWay(id, nodes, tags, metaInformation)
+      }
+
+    def closedWay(nodes: Seq[Long]): Boolean =
+      nodes.headOption.zip(nodes.lastOption).exists { case (head, last) =>
+        head == last
+      }
+
+  }
+
+  final case class Relation(
+      id: Long,
+      relations: Seq[RelationMember],
+      tags: Map[String, String],
+      metaInformation: Option[MetaInformation] = None
+  ) extends OsmEntity {
+    override val osmModel: OsmEntityType = OsmEntityType.Relation
+  }
+
+  object Relation {
+
+    enum RelationMemberType:
+      case Node, Way, Relation, Unrecognized
+
+    final case class RelationMember(
+        id: Long,
```

Comment:
Shouldn't the `OsmEntity` be linked here, like in the previous version?

Reply:
What do you mean by "linked"? As instance of `OsmEntity`? Any reason why we need this?

Reply:
So I think `RelationMember` here is linking to an `OsmEntity` via `id` and `relationTypes`? This would mean that for retrieving the proper `OsmEntity`, one would then need to find it in the sequences of nodes, ways or relations in `OsmContainer`.

Why not setting the reference to the `OsmEntity` here right away, like before? This would mean more effort up front during creation of `OsmContainer`, but quick access every time after that.

Reply:
I think, @sebastian-peter s question is kind of the same type as my question regarding ways. If this is really beneficial in any regard, we should try and note it down somewhere here.

Reply:
Thanks for the clarification. The idea to hold ids only has several justifications:
- The aim of representing osm entities and the osm container _at this point_ is to have our own, simple and lightweight representation of osm data which IMHO should be as close as possible to the way the osm data is provided. As osm data is provided not with references but with ids I prefer having the ids only _at least for this aim_ in PSU.
- osm data from .pbf-files comes unsorted. This means we would enforce a second read of the whole file just to have the node reference, even though we _do not know_ if the user even requires direct node access. In consequence this would at minimum _double_  reading and parsing time for an _not always required_ feature. 

In any way I understand the potential need of having the entity reference directly. However, for this utility class I would prefer the simple way. If there is any situation in the future where a user requires such an extended version of the OsmEntity and the OsmContainer then they might just build their own wrapper class accepting the overhead that comes with such a way of holding the data. If we require something like this in OSMoGrid we might consider adding it in addition to the simple container and entity classes either here or on OSMoGrid itself.

Reply:
ack